force_nevergrad/engine/nevergrad_optimizers.py

```diff
@@ -0,0 +1,215 @@
+
+        # Create optimizer.
+        optimizer = ng.optimizers.registry[self.algorithms](
+            parametrization=instrumentation,
+            budget=self.budget
+        )
+
+        # Minimization/maximization specification
+        minimize_objectives = ['MINI' in k.objective for k in self.kpis]
+
+        # Create a scalar objective nevergrad function from
+        # the MCO function.
+        ng_func = partial(nevergrad_function,
+                          function=func,
+                          is_scalar=True,
+                          minimize_objectives=minimize_objectives
+                          )
+
+        # Optimize.
+        # This returns an nevergrad Instrumentation object.
+        optimization_result = optimizer.minimize(ng_func)
+
+        # Convert the optimal point into MCO format
+        yield translate_ng_to_mco(list(optimization_result.args))
+
+
+@provides(IOptimizer)
+class NevergradMultiOptimizer(HasStrictTraits):
+    """ Optimization of a multi-objective function using nevergrad.
+    """
+
+    #: Algorithms available to work with
+    algorithms = Enum(*ng.optimizers.registry.keys())
+
+    #: Optimization budget defines the allowed number of objective calls
+    budget = PositiveInt(500)
+
+    #: A list of the output KPI parameters representing the objective(s)
+    kpis = List(KPISpecification, visible=False, transient=True)
+
+    def _algorithms_default(self):
+        return "TwoPointsDE"
+
+    def optimize_function(self, func, params):
+        """ Minimize the passed multi-objective function.
+
+        Parameters
+        ----------
+        func: Callable
+            The MCO function to optimize
+            Takes a list of MCO parameter values.
+        params: list of MCOParameter
+            The MCO parameter objects corresponding to the parameters.
+
+        Yields
+        ------
+        list of float or list:
+            The list of parameter values for a single member
+            of the Pareto set.
+        """
+
+        # Create instrumentation.
+        instrumentation = translate_mco_to_ng(params)
+
+        # Create optimizer.
+        optimizer = ng.optimizers.registry[self.algorithms](
+            parametrization=instrumentation,
+            budget=self.budget
+        )
+
+        # Minimization/maximization specification
+        minimize_objectives = ['MINI' in k.objective for k in self.kpis]
+
+        # Create a multi-objective nevergrad function from
+        # the MCO function.
+        ng_func = partial(nevergrad_function,
+                          function=func,
+                          is_scalar=False,
+                          minimize_objectives=minimize_objectives
+                          )
+
+        # Create a MultiobjectiveFunction object from that.
+        ob_func = MultiobjectiveFunction(
+            multiobjective_function=ng_func,
+            upper_bounds=np.array([k.scale_factor for k in self.kpis])
```

Comment:
Could we put in a comment just to note that the `upper_bounds` argument should really come from a `kpi.upper_bound` attribute (which still needs to be implemented), rather than the `scale_factor`? This was documented in the original `NevergradOptimizerEngine` class.

Reply:
What about just leaving it out for now and put it in a new issue (on bdss?)?

Reply:
Either way is fine, as long as it's documented.